proposals/002-text-utf-default.md

```diff
@@ -0,0 +1,156 @@
+started:
+
+-   Modernize the codebase and clear out the bitrot
+
+-   Establish a baseline for performance and any related issues.
+
+-   Update testing and performance benchmarks to make use of `inspection-testing` to ensure fusion is not broken in
+    subsequent UTF-8 related changes.
+
+An MVP should completely preserve standard user-facing API, and not
+break fusion. Performance should not significantly diverge from the
+existing UTF-16 text package. There will be an expected change to the
+exposed Text internals, in which case, breakage should be assessed by
+circulating a git commit reference to a release candidate as soon as
+possible. This candidate should be sourced publicly and loudly.
+
+**Implementation:**
+
+-   TBD: There is a straightforward implementation, but this one is left up to Andrew for comment.
+
+**Stakeholders:**
+
+-   Library authors will need to be made aware of changes and adjust accordingly. HF will provide a git reference to a complete MVP as
+    soon as possible, and produce a migration guide.
+
+-   In the case of GHC, Cabal, and other core infrastructure, we will work closely with these packages to help migrate and assess+fix
+    breakages.
+
+While we do not expect many authors to experience significant changes,
+there will be some help that needs to be given in terms of bumping
+Hackage bounds since this migration will be a major version bump. HF
+will need to coordinate with the Hackage Trustees to help move along
+packages that go out of date.
+
+# Deliverables
+
+-   text-2.0.0.0, which will provide a UTF-8 encoding for Text as a default for all versions going forward.
+
+-   A `text-utf16` package, which is a preservation of the current UTF-16 encoded text, for backwards compatibility.
+
+-   Updates to the Text Haddocks that reflect the UTF-8 changes
+
+-   Announcements and updates across all Haskell channels covering the following:
+    -   Significant dates and milestones
+
+    -   Expected code impact
+
+    -   Release candidates
+
+    -   Delivery
+
+-   Migration instructions and design documentation
+
+# Outcomes
+
+-   Addresses a recognized need and want from both Industry and the Haskell Community
+
+-   Better UX for Haskell's text story
+
+-   Establishes the ability for Haskell Foundation to Get Things Done that were previously blocked for 10+ years in the community.
+
+-   Better interoperability and web story
+
+# Risks
+
+-   HF must minimize the cost to migrate, or people will just get mad (and rightfully so).
+
+-   Text-icu will need a bespoke UTF-8 conversion function. In general, the Unicode story must be tracked and made sure it will not break.
+    -   Recommendation: make this a high-priority deliverable when project planning
+
+-   The old UTF-16 text package will need to be preserved, and will require a maintainer.
+
+-   Performance expectations should be managed: UTF-8 text is \*not\* a panacea. It will not result in a 2x or even significant
```

Comment:
This surprises me -- and in fact calls into question the original motivation.

As I read the proposal as written, I see that this proposal "should completely preserve standard user-facing API" and "not result in ... significant performance increase in many cases", while "performance may regress in some programs". If there is no change to the API and minimal expected change in performance, then what is the concrete benefit here? I do see that people have been clamoring for this change for some time -- that's great. But I do not understand from this proposal what concrete benefit this proposal brings. The best I can extract from this proposal is that it stops people from saying "Haskell is so backward -- it uses UTF-16 as its most common performant text encoding!". Maybe that's the underlying reason? It is good not to be backward, but we should be clear on what our motivation really is, so that we can evaluate later whether we've met our goals.

Reply:
It halves the memory usage in the average case, using 8 instead of 16 for most cases. This can help, or harm performance. It can increase or decrease memory usage. It varies by the workload.

Reply:
I agree with Richard, and think the motivation section is not particularly clear...

> UTF-16 by default requires that all Text values pay a premium for serialization. Arguably, the performance impact of Text is flipped upside-down: most text is UTF-8, and Haskell devs pay an undue cost when working with the wrong default.

This makes complete intuitive sense to me! ...but has this hypothesis been validated? If it's not true then it's not clear what reasons remain for this change.

If measurable performance improvement is what will determine whether this is considered a success and merged (which it sounds like it should), have stakeholders agreed on a set of benchmarks ahead of time?

The other points don't seem like "motivation" to me; e.g. there's no barrier to using `text` with UTF-8 because it's already been written.

Reply:
Can we translate this understanding into a goal of this work? For example, "The geometric mean number of allocations of tests in the xyz benchmark suite goes down by 40%, with no regression above 20%." That would be fantastic! Of course, we might not meet the goal. But @davean is suggesting that the common case would halve memory usage, and that (to me) is in conflict with "not result in ... significant performance increase in many cases". So which is it? The "halve memory usage" seems very plausible -- so we should state that as a goal and then measure to see whether we've met that goal.

Reply:
@goldfirere - There's subtlety in what @davean is saying here: while memory usage may halve in terms of bytes needed to represent text, this does not consider the broader `Text` object (offset + length + array headers), which introduces alignment problems that previously were not there. It's difficult to say whether this will result in an _immediate_ performance increase where say "hey look, here's utf-8!" and suddenly everyone has 2x better memory usage for their text programs. So the distinction is this: 

- We *can* make sure that within the UTF-8 `text` benchmarks we maintain the same fusion guarantees and general asymptotics for its API. 
- We *can't* guarantee it will make Bobby Haskell's immediately 2x faster for all of its text components, because it may fuse or align differently within their program. 
- We *can* guarantee that on the margins, in terms of `text` copying and serialization costs, that memory usage and performance improves. 

So perhaps the distinction needs to be made here in terms of global vs. local benefits. Talking about performance is a minefield unless we set expectations about context. I should reword the point you're citing to reflect that.